OwnTracks work. Beacon logic and testcases

ehagan opened the pull request:
## Description:

OwnTracks handling of iBeacons as fixed region update triggers for HA zones and as tracking tags for devices which cannot track themselves is unreliable. I've experimented with iBeacons in both use cases in several environments and have made these changes to the component. I also rebuilt the test cases to be more representative of what happens in actual use and to more clearly distinguish between the two iBeacon behaviors.


**Related issue (if applicable):** fixes #10181

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#<home-assistant.github.io PR number goes here>

This PR is remade from a previous one which had the wrong email address attached to the commit and which I then made a big mess trying to fix.

## Example entry for `configuration.yaml` (if applicable):
```yaml
mqtt:
  broker: <IP>
  port: 1883
  username: !secret mqtt_username
  password: !secret mqtt_password

# Configure Device Tracker
device_tracker:
  - platform: owntracks
    max_gps_accuracy: 250
    track_new_devices: yes
    waypoints: True
    waypoint_whitelist:
      - <USER>
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

ehagan committed:
OwnTracks work. Beacon logic and testcases

The existing test cases don't really make clear what is being
tested and the iBeacon / Region / Zone / Tracker thing is all
a bit confused.

I'm distinguishing a fixed-place beacon used to trigger entrance
into an HA zone (as a Region Beacon) from a beacon affixed to a
portable or mobile object (as a Mobile Beacon). The behaviors
and test cases for those usages should be different. A Region Beacon
will be named the same as a Home Assistant Zone and seeing an event
from that beacon should trigger a device tracker update related
to that zone. It would be appropriate, though unnecessary, to
configure the Region Beacon with the GPS coordinates of its static
physical location.

A Mobile Beacon is not named after any HA Zone and seeing the beacon
triggers an update in HA setting the location of the beacon to the
current device_tracker location. In this way, when my_phone sees
the beacon on my_keys, the location of my_keys is set to where
my_phone is. And when my_phone stops seeing my_keys, my_keys location
is the location of my_phone the last time it saw them.

A Mobile Beacon's GPS information should be ignored because it's
almost certain to be incorrect because the beacon moves. In fact,
beacons typcially come configured with lat/lon as 0.0/0.0 so using
the location of the beacon in an update has a nasty habit of
setting you and your keys on the bottom of the Atlantic Ocean.

Leave message handling is changed to treat mobile beacons
differently from region beacons and gps regions.

active beacons should be a set. you shouldn't end up
with multiple "active" entries for the same beacon. Let's
enforce that with the correct data structure.

Added test for real-world bug that is fixed.
A series of mobile beacon and region beacon
enter and leave events could cause a mobile
beacon to stick to the tracking device even
though it had tracked through a "leave" event.

Changed two tests to look at the size
of the 'mobile_beacons_active' structure rather
than at the object which will allow this test
to work with any sort of list, set, etc.

homeassistant commented:
Hi @ehagan, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#10183).

Once you do that we will be able to review and accept this pull request.

Thanks!

balloob commented on homeassistant/components/device_tracker/owntracks.py:
```diff
@@ -234,12 +234,36 @@ def async_valid_accuracy(self, message):
+                lon = device_tracker_state.attributes.get("longitude")
+                kwargs['gps_accuracy'] = acc
+                kwargs['gps'] = (lat, lon)
+            except AttributeError:
```
I don't see a case when this will happen?

ehagan replied:
I don't remember why I did this but I suspect it was to catch an issue I had during development. I've been running this code for several days and haven't seen the AttributeError in my logs so I'll remove it.

balloob commented on homeassistant/components/device_tracker/owntracks.py:
```diff
@@ -234,12 +234,36 @@ def async_valid_accuracy(self, message):
+        # changes to kwargs.
+        device_tracker_state = hass.states.get(
+            "device_tracker.{}".format(dev_id))
+        _LOGGER.info("device tracker state: %s", device_tracker_state)
```
Too active logging

balloob commented on homeassistant/components/device_tracker/owntracks.py:
```diff
@@ -234,12 +234,36 @@ def async_valid_accuracy(self, message):
+                _LOGGER.warning('could not get device_tracker state '
+                                'attribute for setting beacon location.')
+
+        # import pdb; pdb.set_trace()
```
dev comment

balloob commented on homeassistant/components/device_tracker/owntracks.py:
```diff
@@ -234,12 +234,36 @@ def async_valid_accuracy(self, message):
+        # kwargs location is the beacon's configured lat/lon
         kwargs.pop('battery', None)
         for beacon in self.mobile_beacons_active[dev_id]:
+            _LOGGER.info("async_see_beacons beacon: %s kwargs %s",
```
We should reduce this logging. level or remove it

balloob commented on homeassistant/components/device_tracker/owntracks.py:
```diff
@@ -297,30 +324,28 @@ def _async_transition_message_leave(hass, context, message, location):
         beacons.remove(location)
         _LOGGER.info("Remove beacon %s", location)
+        yield from context.async_see_beacons(hass, dev_id, kwargs)
+    else:
```
Instead of this `else`, let's add a return to the above `if`. That way we don't get in if…else spaghetti

ehagan replied:
Sounds good.

ehagan committed:
Removing excess logging and unnecessary try catch.

From review on PR #10183 I've removed some info logging
that was unnecessary and I've made the suggested changes
to an if block and a try/catch.

cmsimike commented:
Thank you! I have had some issues with the owntracks/ibeacons functionality where hass seems to get stuck in some sort of loop thinking i've left home, then am home ,then left home, until i restart it.

Hopefully this PR fixes it.

balloob commented:
Thanks! 🐬  🌮
